src/transformers/modeling_encoder_decoder.py

```diff
@@ -18,35 +18,57 @@
 import logging
 import os
 
-from torch import nn
-
 from .modeling_auto import AutoModel, AutoModelWithLMHead
+from .modeling_utils import PreTrainedModel
 
 
 logger = logging.getLogger(__name__)
 
 
-class PreTrainedEncoderDecoder(nn.Module):
+class EncoderDecoderModel(PreTrainedModel):
     r"""
-        :class:`~transformers.PreTrainedEncoderDecoder` is a generic model class that will be
+        :class:`~transformers.EncoderDecoder` is a generic model class that will be
         instantiated as a transformer architecture with one of the base model
-        classes of the library as encoder and (optionally) another one as
+        classes of the library as encoder and another one as
         decoder when created with the `AutoModel.from_pretrained(pretrained_model_name_or_path)`
-        class method.
+        class method for the encoder and `AutoModelWithLMHead.from_pretrained(pretrained_model_name_or_path)` class method for the decoder.
     """
 
     def __init__(self, encoder, decoder):
```

Comment:
So as I detailed below under the `from_pretrained` method, I think here we should allow both an initialization from (pretained) sub-parts and from a config so we can also just reload a saved pretrained model.

Please read the long comment below `from_pretrained` before this one.

Here is a proposal for the `__init__` that should allow both loading:
```python
    def __init__(self, config=None, encoder=None, decoder=None):
        assert config is not None or (encoder is not None and decoder is not None), "Either a configuration or an Encoder and a decoder has to be provided"

        if config is not None:
            self.config = config
            self.encoder = AutoModel.from_config(config.encoder)
            self.decoder = AutoModelWithLMHead.from_config(config.decoder)
        else:
            self.config = EncoderDecoderConfig.from_encoder_decoder_config(encoder.config, decoder.config)
            self.encoder = encoder
            assert (
                self.encoder.get_output_embeddings() is None
            ), "The encoder {} should not have a LM Head. Please use a model without LM Head"
            self.decoder = decoder
```

Reply:
This mean we need an `EncoderDecoderConfig` which will basically be the following:

```python
class EncoderDecoderConfig(PretrainedConfig):
    model_type = "encoder_decoder"
    def __init__(self, **kwargs):
        encoder_config = kwargs.pop('encoder')
        encoder_model_type = encoder_config.pop('model_type')
        decoder_config = kwargs.pop('decoder')
        decoder_model_type = decoder_config.pop('model_type')
        self.encoder = AutoConfig.for_model(encoder_model_type, **encoder_config)
        self.decoder = AutoConfig.for_model(decoder_model_type, **decoder_config)

    @classmethod
    def from_encoder_decoder_config(cls, encoder_config, decoder_config):
        return cls(encoder=encoder_config.to_dict(), decoder=decoder_config.to_dict())

     def to_dict(self):
        """
        Serializes this instance to a Python dictionary. Override the default `to_dict()` from `PretrainedConfig`.

        Returns:
            :obj:`Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance,
        """
        output = copy.deepcopy(self.__dict__)
        output['encoder'] = self.encoder.to_dict()
        output['decoder'] = self.decoder.to_dict()
        output["model_type"] = self.__class__.model_type
        return output
```